Proposal for GPU and PCI passthrough support (#1237)
kubernetes-sigs/cluster-api-provider-vsphere: docs/proposal/20210823-gpu-support.md

#### Story 3 - PCI passthrough for single node Customer

Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.

### Requirements

#### Functional Requirements

* FR1: CAPV MUST support vGPUs.
* FR2: CAPV MUST support GPUDirect.
* FR3: CAPV MUST support PCI Passthrough.

#### Non-Functional Requirements

* NFR1: Unit tests MUST exist for all 3 supports
* NFR2: e2e tests MUST exist for 3 supports

### Implementation Details/Notes/Constraints

#### Controller Changes

```go
// vGPUProfileName is the profileName of the vGPU, in string.
// Defaults to the eponymous property value in the template from which the
// virtual machine is cloned.
// +optional

> VGPUProfileName string `json:"vGPUProfileName,omitempty"`

Reviewer: Missing some info on how this string will get integrated into the existing API types. Also, in order to support PCI passthrough, we'd need more information. From the [docs](https://vdc-download.vmware.com/vmwb-repository/dcr-public/b50dcbbf-051d-4204-a3e7-e1b618c1e384/538cf2ec-b34f-4bae-a332-3820ef9e7773/vim.vm.device.VirtualPCIPassthrough.AllowedDevice.html) the `deviceId` and `vendorId` are required

Reviewer 2: As we are talking about vGPU, GPU direct also, these needs to covered in the schema and possibly in a generic way. What are your thoughts? 
According to the [docs](https://blogs.vmware.com/apps/2018/09/using-gpus-with-virtual-machines-on-vsphere-part-3-installing-the-nvidia-grid-technology.html): 
```
A vGPU profile allows you to assign a GPU solely to one virtual machine’s use or to be used in a shared mode with others.
```
That means `VGPUProfileName ` does not fit with other mode i.e. `pci-passthrough` or `gpu-direct`

Reviewer 3: @sonasingh46 i think we need to define different schema for different techniques, as defined in vm-operator https://github.com/vmware-tanzu/vm-operator/blob/main/config/crd/bases/vmoperator.vmware.com_virtualmachineclasses.yaml#L31-L38
Not sure though, this would be applicable only when each mode has different fields.

Reviewer 2: >i think we need to define different schema for different techniques
Sure.  
I meant is those schemas needs to be present in the CR spec. ( The bigger CR schema ) 
Something like this https://github.com/vmware-tanzu/vm-operator-api/blob/master/api/v1alpha1/virtualmachineclass_types.go#L55

Also, while we are designing the schema, it definitely helps to draw inspiration from the vm-operator api types and it is possible that we might come up with slight different approach for the schema for GPU after discussions. 
Open to discussions.

Author: I have removed 'v' from the variable names as this was not Vgpu specific.

      * [Controller Changes](#controller-changes)
  * [Additional Details](#additional-details)
    * [Test Plan](#test-plan)
    * [Graduation Criteria](#graduation-criteria)
  * [Implementation History](#implementation-history)

## Glossary

* CAPV - An abbreviation of Cluster API Provider vSphere

## Summary

CAPV currently does not support GPU-based infrastructure.

This proposal outlines ways to provision GPU support in CAPV. The proposed changes will maintain backward compatibility and maintain the existing behavior without any extra user configuration.

## Motivation

CAPV does not currently support GPU-accelerated workloads which have a limited addressable market. Competitively, all of the hyper-scale cloud providers (AWS, Azure, & GCP) offer GPU-accelerated virtualization and Kubernetes platforms. Many of them additionally offer robust vertically integrated AI/ML solutions building on open source technology. GPU support is a long-requested feature. With this capability, users can leverage GPU support.

### Goals

1. To enable GPU Support on Vsphere Kubernetes Cluster.
2. To allow vGPU support for multi-node clusters.
3. To allow GPU Direct support for multi-node clusters.

> 4. To allow PCI passthrough support for single-node clusters.

Reviewer: Does PCI passthrough necessarily limit things to a single node cluster? I also don't think goal 1 is needed as 2-4 covers that.

Author: Not sure about it. But there was a discussion going on that we need a Mellanox Network Operator for multi-node.

## Proposal

### User Stories

#### Story 1 - using vGPU

Stacy works at an IT department of an organization where AI/ML models are run every hour and the entire staff wants to leverage GPU-supported nodes. VGPU's are able to provide close to bare-metal performance. Stacy needs a lot of sharing GPUs.

> vGPU provides a full-fledged cluster from which a virtual GPU can be requested.

Reviewer: `vGPU provides a full-fledged cluster from which a virtual GPU can be requested`

I don't really follow this. Isn't vGPU = virtual GPU?

Author: Fixing it.

* FR2: CAPV MUST support GPUDirect.
* FR3: CAPV MUST support PCI Passthrough.

#### Non-Functional Requirements

* NFR1: Unit tests MUST exist for all 3 supports
* NFR2: e2e tests MUST exist for 3 supports

### Implementation Details/Notes/Constraints

#### Proposed Changes

```go
// VirtualMachineCloneSpec is information used to clone a virtual machine.
type VirtualMachineCloneSpec struct {
  // GPUProfileName is the Profile Name of a virtual machine's VGPU, in string.
  // Defaults to the eponymous property value in the template from which the
  // virtual machine is cloned.
  // +optional
  GPUProfileName string `json:"GPUProfileName,omitempty"`
  // GPUDeviceId is the device ID of a virtual machine's GPU, in integer.
  // Defaults to the eponymous property value in the template from which the
  // virtual machine is cloned.
  // +optional
  GPUDeviceId int32 `json:"GPUDeviceId,omitempty"`
  // GPUVendorId is the vendor ID of a virtual machine's GPU, in integer.
  // Defaults to the eponymous property value in the template from which the
  // virtual machine is cloned.
  // +optional
  GPUVendorId int32 `json:"GPUVendorId,omitempty"`
}
```

#### Controller Changes

> * If VGPU or PCI device is present, the CRD is fetched and used to identify the expected devices.

Reviewer: which CRD will get fetched here?

Author: I have rephrased this  since no new CRD would be fetched and existing CRD's would be used.

* The controller initiallizes variables respective to VGPU i.e `GPUProfileName`.
* Similarly, controller also initializes variables related to Direct devices(for PCI passthrough and GPUDirect) ie. `GPUDeviceId` and `GPUVendorId`.
* Using the above variables, respective objects of backing info of type `vimTypes.VirtualPCIPassthroughVmiopBackingInfo` with `{Vgpu: <GPUProfileName>}` for VGPU's and `vimTypes.VirtualPCIPassthroughAllowedDevice` for GPUDirect should be created by the controller.

> * Controller will call a function called as createPCI device which will use the information in above step to create objects of type `vimTypes.VirtualPCIPassthrough`

Reviewer: should the VirtualMachineClone spec contain an array of PCI devices?

Author: I have added the fields into a GPU struct and loaded it inside of VirtualMachineClone as a slice.

# GPU support CAPV

```text
---
title: GPU support CAPV
authors:
  - "@geetikabatra"
reviewers:
  - "@vijaykumar"
  - "@ankitaswamy"

>   - "@ashutoshkumar"

Reviewer 2: If you want to put it, my github ID is : sonasingh46
Hence,
```
@ashutoshkumar ---> @sonasingh46
```

  * [Motivation](#motivation)
    * [Goals](#goals)
  * [Proposal](#proposal)
    * [User Stories](#user-stories)
      * [Story 1 - using vGPU](#story-1---using-vgpu)
      * [Story 2 - GPU Direct implementation](#story-2---gpu-direct-implementation)
      * [Story 3 - PCI passthrough for single node Customer](#story-3---pci-passthrough-for-single-node-customer)
    * [Requirements](#requirements)
      * [Functional Requirements](#functional-requirements)
      * [Non-Functional Requirements](#non-functional-requirements)
    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
      * [Controller Changes](#controller-changes)
  * [Additional Details](#additional-details)
    * [Test Plan](#test-plan)
    * [Graduation Criteria](#graduation-criteria)
  * [Implementation History](#implementation-history)

## Glossary

* CAPV - An abbreviation of Cluster API Provider vSphere

## Summary

CAPV currently does not support GPU-based infrastructure.

This proposal outlines ways to provision GPU support in CAPV. The proposed changes will maintain backward compatibility and maintain the existing behavior without any extra user configuration.

> 

Reviewer 2: n_p:
To make these lines more clear, I was thinking to put some lines explaining what does GPU support in CAPV means?

e.g. 

```
CAPV currently does not support provisioning of GPU-based worker nodes and hence limits the deployment of workloads that needs GPU 

This proposal outlines ways to enable GPU support in CAPV. This means that the k8s cluster created via CAPV can have GPU enabled nodes that can be finally consumed by the workloads(pods) that request GPU.   The proposed changes will maintain backward compatibility and maintain the existing behaviour without any extra user configuration.

"As an end user, I want to deploy GPU accelerated workloads on the Kubernetes cluster created via CAPV

```

Author: @sonasingh46. I think it is self-explanatory.  But, if you strongly feel I should rephrase it. I would certainly do that.

Reviewer 2: I was only confused by `provision GPU support in CAPV`. It is just a nit pick from me. Let us see what others have to say. Should be fine.

## Motivation

CAPV does not currently support GPU-accelerated workloads which have a limited addressable market. Competitively, all of the hyper-scale cloud providers (AWS, Azure, & GCP) offer GPU-accelerated virtualization and Kubernetes platforms. Many of them additionally offer robust vertically integrated AI/ML solutions building on open source technology. GPU support is a long-requested feature. With this capability, users can leverage GPU support.

### Goals

1. To enable GPU Support on Vsphere Kubernetes Cluster.
2. To allow vGPU support for multi-node clusters.
3. To allow GPU Direct support for multi-node clusters.
4. To allow PCI passthrough support for single-node clusters.

## Proposal

### User Stories

#### Story 1 - using vGPU

Stacy works at an IT department of an organization where AI/ML models are run every hour and the entire staff wants to leverage GPU-supported nodes. VGPU's are able to provide close to bare-metal performance. Stacy needs a lot of sharing GPUs.

> vGPU provides a full-fleshed cluster from which a virtual GPU can be requested.

Reviewer 2: Requesting a vGPU does not necessarily mean that it can be shared across, right? 
One can still request a vGPU and it may not be shareable.  
For example, in the below doc ( specific to Nvidia ), I can see that we can request vGPU in multiple modes and one of the mode can allow vGPU to be shared. 
https://docs.nvidia.com/grid/latest/grid-vgpu-user-guide/index.html

The question I am trying to ask is that, when we say that we have a vGPU available, does it inherently mean that it is shareable? 
Are we trying to say that from CAPV end user prespective, when a user request for a vGPU that always means it is shareable ( we make sure from out implementation that we provision vGPU in a shareable mode ), which sounds fine to me.

> vGPU provides a full-fleshed cluster from which a virtual GPU can be requested.

Author: Inherently it's not shareable. Sharing is enabled inside of physical hardware using one of the above techniques you pointed out, which is the user's own choice. We assume it to be just a piece of GPU(shareable or non-shareable).  Sharing should be automatically handled by the vGPU instance(Need to double-check that).

#### Story 2 - GPU Direct implementation

> Tony is an Engineer at a big organization with multiple centers across the Globe. Tony needs a big GPU cluster to address the needs of his organization. So they intend to use GPU Direct. GPU direct is technically a pool of GPUs connected by a network card in a kind of peer-to-peer network. This is the latest technology and gives the advantage of shared GPUs from a pool. GPU  direct is the technology that helps in linking multiple GPUs. Tony wants to leverage this technology so that it is easier for him to manage resource allocation.

Reviewer 2: vGPU and GPU direct seems to be available by Nvidia. Is this vendor dependent? Are there other providers also providing this tech/feature? 
These questions may help design the CR spec schema in my opinion.

Author: This is indeed vendor-dependent.  This proposal is keeping in mind Nvidia GPUs and it's specific operator.  Intel has an entirely different architecture and method. Currently, we are targeting Nvidia.

Reviewer 4: Since this implementation is vendor dependent, we should add a section around saying this implementation is specific to Nvidia.

#### Story 3 - PCI passthrough for single node Customer

> Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.

Reviewer 2: A non-goal item but pasting here for reference. 
https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/#node-labeller

Reviewer 2: >The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.
Trying to understand this part. What is meant by node migration here. 

I am thinking the following: 
If a node for example, `node-1` has GPU available via PCI passthrough and for some reason this node cannot be used, can't we bring up a new node `node-1'` and make a GPU available to it and then de-comission the previous `node-1`. 
The workloads should actually get migrated automatically to the newer nodes as we will have the same  GPU label

### Requirements

#### Functional Requirements

* FR1: CAPV MUST support vGPUs.
* FR2: CAPV MUST support GPUDirect.
* FR3: CAPV MUST support PCI Passthrough.

#### Non-Functional Requirements

* NFR1: Unit tests MUST exist for all 3 supports
* NFR2: e2e tests MUST exist for 3 supports

### Implementation Details/Notes/Constraints

#### Controller Changes

```go
// vGPUProfileName is the Profile Name of a virtual machine's vGPU, in string.
// Defaults to the eponymous property value in the template from which the
// virtual machine is cloned.
// +optional
VGPUProfileName string `json:"vGPUProfileName,omitempty"`
// vGPUDeviceId is the device ID of a virtual machine's vGPU, in integer.
// Defaults to the eponymous property value in the template from which the
// virtual machine is cloned.

> // +optional

Reviewer 5: The optional fields of revisionId, subDeviceId, subVendorId should also be included

> // +optional

Author: The keyword optional here implies that the node might or might not have GPU attached.
Also, the implementation ignores the fields `revisionId, subDeviceId, subVendorId`. This is debatable though. I am open to discussion regarding this.

VGPUDeviceId int32 `json:"vGPUDeviceId,omitempty"`
// vGPUVendorId is the vendor ID of a virtual machine's vGPU, in integer.
// Defaults to the eponymous property value in the template from which the
// virtual machine is cloned.
// +optional
VGPUVendorId int32 `json:"vGPUVendorId,omitempty"`
```

```yaml
vGPUDeviceId:
  description: vGPUDeviceId is the device ID of a virtual machine's
    vGPU, in integer. Defaults to the eponymous property value in
    the template from which the virtual machine is cloned.
  format: int32
  type: integer
vGPUProfileName:
  description: vGPUProfileName is the Profile Name of a virtual
    machine's vGPU, in string. Defaults to the eponymous property
    value in the template from which the virtual machine is cloned.
  type: string
vGPUVendorId:
  description: vGPUVendorId is the vendor ID of a virtual machine's
    vGPU, in integer. Defaults to the eponymous property value in
    the template from which the virtual machine is cloned.
  format: int32
  type: integer
```

## Additional Details

### Test Plan

> * Unit tests for cluster controller to test behavior vGPU is requested.

Reviewer 5: ```suggestion
* Unit tests for cluster controller to test behavior when vGPU is requested.
```

  // GPU is the GPU configuration for this machine's VM.
 // +optional
 Gpu GpuSpec `json:"gpu,omitempty"`
}

//GPU defines virtual machine's GPU configuration
type GpuSpec struct {
 // Devices is the list of gpu devices used by the virtual machine.
 Devices []GpuDeviceSpec `json:"devices"`
}

//GPUDeviceSpec defines virtual machine's GPU configuration
type GpuDeviceSpec struct {
 // GPUProfileName is the Profile Name of a virtual machine's VGPU, in string.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 GPUProfileName string `json:"GPUProfileName,omitempty"`

 // GPUDeviceId is the device ID of a virtual machine's GPU, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 GPUDeviceId int32 `json:"GPUDeviceId,omitempty"`

 // GPUVendorId is the vendor ID of a virtual machine's GPU, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 GPUVendorId int32 `json:"GPUVendorId,omitempty"`
}

```

#### Controller Changes

> - If vGPU is present, the CR populates the ProfileName, which adds the vGPU to the list of devices.

Author: @srm09 PTAL

> # GPU support in CAPV

Reviewer 4: We had discussed updating the title of the proposal to explicitly call out the PCI Passthrough support, could you include that here?

Author: Had updated an outdated version in my previous commit. Updated the new one.

```text
---

> title: GPU support in CAPV

Reviewer 4: Same as above

last-updated: 2021-08-25
status: implementable
```

## Table of Contents

- [GPU support in CAPV](#gpu-support-in-capv)
  - [Table of Contents](#table-of-contents)
  - [Glossary](#glossary)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
  - [Proposal](#proposal)
    - [User Stories](#user-stories)
      - [Story 1 - using vGPU](#story-1---using-vgpu)
      - [Story 2 - GPU Direct implementation](#story-2---gpu-direct-implementation)
      - [Story 3 - PCI passthrough for single node Customer](#story-3---pci-passthrough-for-single-node-customer)
    - [Requirements](#requirements)
      - [Functional Requirements](#functional-requirements)
      - [Non-Functional Requirements](#non-functional-requirements)
    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
      - [Proposed Changes](#proposed-changes)
      - [Controller Changes](#controller-changes)
  - [Additional Details](#additional-details)
    - [Test Plan](#test-plan)
    - [Graduation Criteria](#graduation-criteria)
  - [Implementation History](#implementation-history)

## Glossary

- CAPV - An abbreviation of Cluster API Provider vSphere

## Summary

> CAPV currently does not support GPU-based infrastructure.

Reviewer 4: nit: remove this line, the below paragraph talks about the same thing.

> CAPV currently does not support provisioning of GPU-based worker nodes and hence limits the deployment of workloads that needs GPU. This proposal outlines ways to enable GPU support in CAPV. This means that the k8s cluster created via CAPV can have GPU enabled nodes that can be finally consumed by the workloads(pods) that request GPU. The proposed changes will maintain backward compatibility and maintain the existing behaviour without any extra user configuration. "As an end user, I want to deploy GPU accelerated workloads on the Kubernetes cluster created via CAPV

Reviewer 4: ```suggestion
CAPV currently does not support provisioning of GPU-based worker nodes and hence limits the deployment of workloads that needs GPU. This proposal outlines ways to enable GPU support in CAPV. This means that the k8s cluster created via CAPV can have GPU enabled nodes that can be finally consumed by the workloads(pods) that request GPU. The proposed changes will maintain backward compatibility and maintain the existing behaviour without any extra user configuration.
```

## Motivation

> CAPV does not currently support GPU-accelerated workloads which have a limited addressable market. Competitively, all of the hyper-scale cloud providers (AWS, Azure, & GCP) offer GPU-accelerated virtualization and Kubernetes platforms. Many of them additionally offer robust vertically integrated AI/ML solutions building on open source technology. GPU support is a long-requested feature. With this capability, users can leverage GPU support.

Reviewer 4: ```suggestion
CAPV does not currently support GPU-accelerated workloads which have a limited addressable market. Competitively, all of the hyper-scale cloud providers (AWS, Azure, & GCP) offer GPU-accelerated virtualization and Kubernetes platforms. Many of them additionally offer robust vertically integrated AI/ML solutions building on open source technology. GPU support is a long-requested feature. With this capability, users can deploy GPU accelerated workloads on the Kubernetes cluster provisioned via CAPV
```

#### Story 3 - PCI passthrough for single node Customer

Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.

### Requirements

#### Functional Requirements

- FR1: CAPV MUST support vGPUs.
- FR2: CAPV MUST support GPUDirect.
- FR3: CAPV MUST support PCI Passthrough.

#### Non-Functional Requirements

- NFR1: Unit tests MUST exist for all 3 supports

> ### Implementation Details/Notes/Constraints

Reviewer 4: Add a section for **Constraints** and call out the discussion in the [comment](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1237#discussion_r813153403) around placement of VMs when using Failure Domains.

Author: Done!

#### Proposed Changes

```go

// VirtualMachineCloneSpec is information used to clone a virtual machine.
type VirtualMachineCloneSpec struct {

>   // GPU is the GPU configuration for this machine's VM.

Reviewer 4: nit
```suggestion
  // GPU is the GPU configuration for the virtual machine.
```

}

```

#### Controller Changes

- If vGPU is present, the CR populates the ProfileName, which adds the vGPU to the list of devices.
- If GPU is present via PCI passthrough, `device ID`, and `vendor ID`. The CR populates the variables respective to VGPU, i.e., GPUProfileName.
- Similarly, the controller also initializes variables related to Direct devices(for PCI passthrough and GPUDirect). i.e., `GPUDeviceId` and `GPUVendorId`.
- Controller will call a function `getGpuSpecs` device, which will use the information stated above to determine if that specific GPU is present or not. If it is, then VM with that GPU is spawned.

## Additional Details

### Test Plan

- Unit tests for cluster controller to test behaviour when vGPU is requested.
- Unit tests for cluster controller to test behaviour if vGPU works as expected.
- Unit tests for cluster controller to test if multiple vGPUs are requested, do all of them behave as expected.
- Unit tests for cluster controller to test GPU's requested from GPU direct work as expected.
- Unit tests for cluster controller to test PCI passthrough works as expected.
- Unit tests for a single node PCI passthrough.

> - Enabling GPU based infrastructure.
> - Setting up webhooks to trigger the GPU based E2E tests.
> - E2E test to create a GPU-enabled cluster with one control plane node and one worker node.
>   - Subtasks:
>     - Wait for a node to have a "nvidia.com/gpu" allocatable resource.
> - E2E test to run a GPU-based calculation.
>   - Run a CUDA vector calculation job

Reviewer 4: ```suggestion
- E2E tests for GPU enabled workload clusters
   #### Subtasks:
   - Enabling GPU based infrastructure, will be hosted behind a VPN.
   - Setting up webhooks to trigger the GPU based E2E tests.
   - Actual test scenarios
      - E2E test to create a GPU-enabled cluster with one control plane node and one worker node.
      - Wait for a node to have a "nvidia.com/gpu" allocatable resource.
      - E2E test to run a GPU-based calculation.
      - Run a CUDA vector calculation job
```

#### Story 3 - PCI passthrough for single node Customer

Alex is an Engineer at a retail organization that requires a single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU-connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. The challenge that Alex can face is that using passthrough Alex wouldn't be able to migrate nodes.

### Requirements

#### Functional Requirements

- FR1: CAPV MUST support vGPUs.
- FR2: CAPV MUST support GPUDirect.

- FR3: CAPV MUST support PCI Passthrough.

#### Non-Functional Requirements

- NFR1: Unit tests MUST exist for all 3 supports
- NFR2: e2e tests MUST exist for 3 supports

### Implementation Details/Notes/Constraints

#### Proposed Changes

```go

// VirtualMachineCloneSpec is information used to clone a virtual machine.
type VirtualMachineCloneSpec struct {
  // Vgpu is the Vgpu configuration for this machine's VM.
 // +optional

>  Vgpu []VGPUSpec `json:"vgpu,omitempty"`

Reviewer 6: why don't we instead have something like this?
```golang
VGPUDevices []VGPUDeviceSpec `json:"vgpuDevices,omitempty"` 
```

Author: Done!

 // Pci is the Pci configuration for this machine's VM.
 // +optional

>  Pci []PCISpec `json:"pci,omitempty"`

Reviewer 6: why don't we instead have something like this?
```golang
PCIDevices []PCIDeviceSpec `json:"pciDevices,omitempty"`
```

Reviewer 4: +1 to what Yassine has said here ^

Author: Done!

}

//PCISpec defines virtual machine's PCI configuration

> type PCISpec struct {

Reviewer 6: I don't think we need this struct here

Author: Removed!

 // Devices is the list of pci devices used by the virtual machine.
 PciDevices []PCIDeviceSpec `json:"pcidevices"`
}

//VGPUSpec defines virtual machine's VGPU configuration

> type VGPUSpec struct {

Reviewer 6: I don't think we need this struct

Reviewer 4: The VGPUSpec adds nothing apart from a wrapper around the VGPUDeviceSpec. Makes sense to remove this too. The NetworkSpec comprises of a bunch of other fields which are a part of the Spec, hence having the wrapper struct makes more sense in that case. And this approach applies to all the other structs above.

Author: Done!

 // Devices is the list of vgpu devices used by the virtual machine.
 VgpuDevices []VGPUDeviceSpec `json:"vgpudevices"`
}

//VGPUDeviceSpec defines virtual machine's VGPU configuration
type VGPUDeviceSpec struct {
 // ProfileName is the Profile Name of a virtual machine's VGPU, in string.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 ProfileName string `json:"profileName,omitempty"`
}

//PCIDeviceSpec defines virtual machine's PCI configuration
type PCIDeviceSpec struct {

 // DeviceId is the device ID of a virtual machine's PCI, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional

>  DeviceId int32 `json:"deviceId,omitempty"`

Reviewer 6: nit: `DeviceID`

 // VendorId is the vendor ID of a virtual machine's PCI, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional

>  VendorId int32 `json:"vendorId,omitempty"`

Reviewer 6: nit: `VendorID`

}

> ```

Author: Done!

> ```

Reviewer 4: @geetikabatra do not see the validations listed out in the proposal

> #### Controller Changes

Reviewer 4: From my POV, it would make more sense to the reader to list out the changes required for each GPU support type.
For example:
for vGPU, populate the ProfileName and add to the list of devices.
for PCI passthrough, populate the device and vendor IDs and add to the list of devices.
and so on.

> #### Controller Changes

Author: @srm09 I added a detailed overview of the work as Gab had asked me to do so a couple of months back. I can remove it if you want.

> #### Controller Changes

Author: /rephrase

- If VGPU is present, VGPU's `ProfileName` would be fetched from `VirtualMachineCloneSpec` and device object (`createPCIPassThroughDevice` from govmomi library) is formed using this VGPU `ProfileName`. This device object is added to

> - the slice of devices and then used by `vspheremachine_controller` to schedule nodes.

Reviewer 4: The VM placement is dictated by the failureDomain present on the VirtualMachineSpec (maybe placement on a particular host or group op hosts), so what happens if the GPUProfile is not available on that host.

Author: Need a broader discussion on this.

Author: If we are using the Vsphere failure domain to dictate where the sphere VM gets created. It needs to have access to GPU, otherwise cloning will fail. It is users' responsibility to make sure that whatever failure domain they are trying to put the VM on(GPU VM). It must have the access to the required physical device.

- If PCI device is present, PCI device's `DeviceId` and `VendorId` would be fetched from `VirtualMachineCloneSpec` and device object(`createPCIPassThroughDevice` from govmomi library) is formed using this information and then used by the `vspheremachine_controller` to schedule nodes.

> - The controller initiallizes variables respective to VGPU i.e `GPUProfileName`.

Author: It basically is another way of saying that `GPUProfile name` is now set and can be used.

- Using the above variables, respective objects of backing info of type `vimTypes.VirtualPCIPassthroughVmiopBackingInfo` with `{Vgpu: <GPUProfileName>}` for VGPU's and `vimTypes.VirtualPCIPassthroughAllowedDevice` for GPUDirect should be created by the controller.
- Controller will call a function called as createPCI device which will use the information in above step to create objects of type `vimTypes.VirtualPCIPassthrough`
- The create PCI device function will also be responsible for comparing the current devices with the expected devices.
- The create method will return a slice `deviceChanges` of type `vimTypes.BaseVirtualDeviceConfigSpec`.
- If there are no current devices and new devices are added. The slice `devicechanges` would reflect only add operations which would be equivalent to the number of devices added.
- If the current devices are same as proposed devices. The length of the device changes remains 0.
- Adding Vgpu devices and PCI devices with differnet backing info would return one remove and one add operation for each device.

> So the length of device changes would be 2*devices with different backing info.

Reviewer 4: Since the VSphereMachineTemplate and VSphereMachine objects are immutable, how would this case arise? Why would vGPU or PCI devices with different backing info be needed?

Author: Just curious, how does Network Devices work. What I see is, they call a `Finder.Network` in `govmomi` and just return that list of devices. In the case of GPU's there is no such method as `finder.PCIDevices`. All the workflow above will not be required if we have a list of GPU devices. 

P.S If above stands true, then I guess, I have to implement a PCI finder in govmomi.

Author: //need to file an issue in govmomi

Reviewer 4: This seems to be taken care of by https://github.com/vmware/govmomi/pull/2772. Resolving this

## Additional Details

### Test Plan

- Unit tests for cluster controller to test behavior when vGPU is requested.
- Unit tests for cluster controller to test behavior if vGPU works as expected.
- Unit tests for cluster controller to test if multiple vGPU's are requested, do all of them behave as expected.
- Unit tests for cluster controller to test GPU's requested from GPU direct work as expected.
- Unit tests for cluster controller to test PCI passthrough works as expected.

> - E2E test with a GPU Direct cluster running in different availability zones.

Reviewer 4: Upstream CI does not have GPU support, so running E2E tests would not be possible.

Author: Need a broader discussion on this.

Reviewer 4: Lets create an issue to track the E2E testing for GPU separately. Whatever mechanism we end up taking to run the tests on GPU specialized hardware hosted behind a VPN, we should use the issue to describe the solution and update the progress.

Author: Noted this one as a work item. Marking this resolved.

- Unit tests for a single node PCI passthrough.

> - E2E test with the PCI passthrough.

Reviewer 4: same as above

 PciDevices []PCIDeviceSpec `json:"pcidevices"`
}
//VGPUSpec defines virtual machine's VGPU configuration
type VGPUSpec struct {
 // Devices is the list of vgpu devices used by the virtual machine.
 VgpuDevices []VGPUDeviceSpec `json:"vgpudevices"`
}
//VGPUDeviceSpec defines virtual machine's VGPU configuration
type VGPUDeviceSpec struct {
 // ProfileName is the Profile Name of a virtual machine's VGPU, in string.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 ProfileName string `json:"profileName,omitempty"`
}
//PCIDeviceSpec defines virtual machine's PCI configuration
type PCIDeviceSpec struct {
 // DeviceID is the device ID of a virtual machine's PCI, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 DeviceID int32 `json:"deviceId,omitempty"`
 // VendorId is the vendor ID of a virtual machine's PCI, in integer.
 // Defaults to the eponymous property value in the template from which the
 // virtual machine is cloned.
 // +optional
 VendorID int32 `json:"vendorId,omitempty"`
}
```

#### Current Scope of vendor

- Nvidia

> #### Additional Webhooks

Author: @srm09 This section says all about validations.

2. To allow vGPU support for multi node clusters
3. To allow GPU Direct support for multi node clusters
4. To allow PCI passthrough support for single node clusters.

## Proposal

### User Stories

#### Story 1 - using vGPU

Stacy works at an IT department of a organization where AI/ML models are run every hour and the entire staff wants to leverage GPU supported nodes. VGPU's are able to provide close to bare metal performance. Stacy needs a lot of sharing GPU’s. 
vGPU provides a full fleshed cluster from which a virtual GPU can be requested.

#### Story 2 - GPU Direct implementation

Tony is an Engineer at a big organization with multiple centers across the Globe. Tony needs a big GPU cluster to address the needs of his 
organization. So they intend to use GPU Direc. GPU direct is  technically a pool of GPU’s connected by a network card in a kind of  peer to peer network. This is the latest technology and gives advantage of shared GPU’s from a pool. GPU  direct is the technology which helps in linking multiple GPU’s. Tony wants to leverage this technology so that it is easier for him to manage resource allocation.

#### Story 3 - PCI passthrough for single node Customer

> Alex is an Engineer at retail organization that requires single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. Challenges that Alex can face is that using passthrough Alex wouln't be able to migrate nodes.

Author: @MaxRink I don't get your question here. Non GPU will be counted among regular nodes in my opinion.

> Alex is an Engineer at retail organization that requires single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. Challenges that Alex can face is that using passthrough Alex wouln't be able to migrate nodes.

Reviewer 7: I mean passthrough for PCI-E devices that are not GPUs, as PCI-E passthrough shouldnt really care if its a GPU, a NIC or even an HBA

> Alex is an Engineer at retail organization that requires single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. Challenges that Alex can face is that using passthrough Alex wouln't be able to migrate nodes.

Author: Yeah, If we put it that way, yes. It also applies for non GPU devices.

> Alex is an Engineer at retail organization that requires single GPU node. They use one node with GPU attached and want to keep things simple. Alex can simply add this GPU connected machine to the cluster and that should do the job. While selecting nodes, Alex can use appropriate labels to run his AI/ML workload on this particular node. PCI passthrough will provide direct GPU support. Challenges that Alex can face is that using passthrough Alex wouln't be able to migrate nodes.

Reviewer 4: @yastij Does it make sense to separate out the PCI Passthrough support to a separate proposal? This could also enable the workflow of provisioning VMs physical NICs attached to the VMs created by CAPV.

And if it is too much work to separate the proposal out, I +1 Naadir's comments on retitling the proposal since PCI passthrough support is not strictly GPU related.
